File: DotNet/Finance/FromFileTransaction.cs

using System;  
using System.IO;  // Import the File class
using System.Collections.Generic; // Import class to deal with enumerators
using System.Globalization;



namespace Finance {

// This class is for a PEX business service that reads a transaction request message from a file. It then parses the 
// file format of that message and converts it into a Finance.TransactionRequest message object, before sending it to
// another business component declared at runtime in the property TargetComponentNamespublic class FromFileTransaction : InterSystems.EnsLib.PEX.BusinessService {
public class FromFileTransaction : InterSystems.EnsLib.PEX.BusinessService {
    public string TargetComponentNames; // Comma separated list of target production components. Set at runtime.


    public override void OnTearDown() {} // Abstract method in PEX superclass. Must override.
    public override void OnInit() { } // Abstract method in PEX superclass. Must override.

    // OnProcessInput is called at an interval specified in the 'call interval' setting in the production.
    public override object OnProcessInput(object input)  {
        
            // Cast input to string representing file path
            string file = (string)input;
            
            // Create enumerable for lines of file.
            IEnumerable<string> FileLines = File.ReadLines(file);
            IEnumerator<string> reader = FileLines.GetEnumerator();

            // Instantiate TransactionRequest message object, passing the lines of file into a constructor function.
            TransactionRequest request = new TransactionRequest(reader);

            // Delete file after reading data.
            File.Delete(file);

            // Iterate through target business components and send request message
            string[] targetNames =  TargetComponentNames.Split(",");
            foreach (string name in targetNames){
              SendRequestAsync(name, request);
              
            }

        
        return null;
    }
} 

}

File: Java/src/Finance/FromFileTransaction.java

package Finance;
import java.io.File;  // Import the File class.
import java.io.FileNotFoundException;  // Import this class to handle errors.
import java.util.Scanner; // Import the Scanner class to read text files.
import Finance.TransactionRequest; // Import the Transaction message to send through production.



// This class is for a PEX business service that reads a transaction request message from a file. It then parses the 
// file format of that message and converts it into a Finance.TransactionRequest message object, before sending it to
// another business component declared at runtime in the property TargetComponentNames
public class FromFileTransaction extends com.intersystems.enslib.pex.BusinessService {
    public String TargetComponentNames; // Comma separated list of target production components. Set at runtime.
    

    public void OnTearDown() {} // Abstract method in PEX superclass. Must override.
    public void OnInit() {} // Abstract method in PEX superclass. Must override.

    // OnProcessInput is called at an interval specified in the 'call interval' setting in the production.
    public java.lang.Object OnProcessInput(java.lang.Object messageInput) throws java.lang.Exception {

      String path = (String)messageInput;
      
      File file = new File(path);

      
      // Create scanner object to read file lines.
      Scanner reader = new Scanner(file);

      // Instantiate TransactionRequest message object using constructor, passing in each line as an argument.
      TransactionRequest request = new TransactionRequest(reader.nextLine(),reader.nextLine(),reader.nextLine(),reader.nextLine(),reader.nextLine());

      reader.close();

      // Delete file after processing.
      file.delete();

      // Split target business component string and send to each component.
      String[] targetNames =  TargetComponentNames.split(",");
      for (String name : targetNames){
        SendRequestAsync(name, request);
      }
    
      return null;
    }
}